### ISSUE TITLE
NET 6 Application crashes without a StackTrace (I've a MiniDump) Code: 0xc0000374

### ISSUE THREAD

**Issue by jogibear9988** [dotnet/runtime#76875]
### Description

In the Dump I see this Exception:

Ausnahmefehler bei 0x00007FF84B45C259 (ntdll.dll) in MCC.Common.Services.NTServiceHost.exe.4844.dmp: 0xC0000374: Ein Heap wurde beschädigt (Parameter: 0x00007FF84B4C37F0)

![image](https://user-images.githubusercontent.com/364896/195102985-5ada87ac-f666-4a88-9c5d-4904b0c03af5.png)

The MiniDump is here:
https://owncloud.mlog-onlineportal.de/index.php/s/Lv5LSfsIUTywKLC
Password: mlog

We got this Message in the EventLog when our C# Program crashed on customer Side:

Name der fehlerhaften Anwendung: MCC.Common.Services.NTServiceHost.exe, Version: 6.0.0.0, Zeitstempel: 0x62fff1bf
Name des fehlerhaften Moduls: ntdll.dll, Version: 10.0.17763.3469, Zeitstempel: 0x12c01028
Ausnahmecode: 0xc0000374
Fehleroffset: 0x00000000000fc259
ID des fehlerhaften Prozesses: 0x12ec
Startzeit der fehlerhaften Anwendung: 0x01d8dd00c5020154
Pfad der fehlerhaften Anwendung: C:\MLOG\MCC\application\service\MCC.Common.Services.NTServiceHost.exe
Pfad des fehlerhaften Moduls: C:\Windows\SYSTEM32\ntdll.dll
Berichtskennung: 932dcffc-8a05-4ad2-9723-25a013a8e0aa
Vollständiger Name des fehlerhaften Pakets:
Anwendungs-ID, die relativ zum fehlerhaften Paket ist:

### Reproduction Steps

not reproducable

### Expected behavior

Program should not crash

### Actual behavior

Crashed once

**Comment by jkotas** [comment]
Windows exception code 0xC0000374 is `STATUS_HEAP_CORRUPTION`. The status codes are documented here: https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-erref/596a1078-e883-4972-9bbc-49e60bebca55.

Thank you for sharing the crashdump. The crash is likely caused by a bug in the following printer driver that is loaded in the process:
```
        CompanyName:      Euro Plus d.o.o.
        ProductName:      Thermal Printers Driver
        InternalName:     ZDNUI56.DLL
        OriginalFilename: ZDNUI56.DLL
        ProductVersion:   5.1.17.7644
        FileVersion:      5.1.17.7644 built by: WinDDK
        FileDescription:  Thermal WIN2K/XP Printer Driver User Interface
        LegalCopyright:   Copyright (C) 1997..2022 Euro Plus d.o.o.
```

You can try following one of tutorials out there for debugging unmanaged heap corruptions (e.g. by enabling pageheap) or ask the printer driver vendor for help.

Also, I have noticed that the app is running on .NET Runtime 6.0.3. You may want to update to .NET 6.0.9 that is the latest servicing update (https://dotnet.microsoft.com/en-us/download/dotnet/6.0) that includes reliability fixes.

**Comment by jogibear9988** [comment]
@jkotas thought of something like this, but thanks for clarifying this. We look to contact printer driver manufacturer.

But should the exception type be logged? (Issue #76741 )

**Comment by jkotas** [comment]
There are many different Windows exception status codes. Most of them do not have managed exception equivalents. For example, there is no HeapCorruptedException that would match `STATUS_HEAP_CORRUPTION`.

We could copy & paste the doc page that I have shared, reformat it as a mapping table and use it to convert the exception code into text. I am not sure whether it is worth it. We do not have a mapping table like this build into the runtime for any other types of errors, and searching the internet with the status code seems to work about as good as searching with status code name.

**Comment by codinghardadmin** [comment]
@jkotas 
Our customer changed the driver version. The same error occurred here. 
Here is the new dump: https://owncloud.mlog-onlineportal.de/index.php/s/UY7mpA9oCDjeWBt
Password: mlog

Could you tell us the printer driver version? Or how can I get the driver version myself? Where do you see that the crash is from the printer driver? 

Thanks in advance

**Comment by jkotas** [comment]
> Or how can I get the driver version myself?

1. Open the dump in windbg (windbg comes as part [Debugging Tools for Windows](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools)
2. Run `lm` command. It will list all modules loaded in the process. These printer driver modules are the first few in the list with cryptic names like `ZDNdrv56`
3. Run `lm v m<driver module name>`, for example `lm v mZDNdrv56`. It will print the verbose version information from the module.

From the last you have shared:
```
0:017> lm v mZDNdrv56
Browse full module list
start             end                 module name
00000000`00400000 00000000`00578000   ZDNdrv56   (deferred)             
    Image path: C:\Windows\System32\spool\drivers\x64\3\ZDNdrv56.dll
    Image name: ZDNdrv56.dll
    Browse all global symbols  functions  data
    Timestamp:        Fri Jul 28 04:05:35 2017 (597B1A7F)
    CheckSum:         0019EC43
    ImageSize:        00178000
    File version:     0.3.0.0
    Product version:  5.1.16.6816
    File flags:       8 (Mask 3F) Private
    File OS:          40004 NT Win32
    File type:        2.1 Dll
    File date:        00000000.00000000
    Translations:     0409.04b0
    Information from resource tables:
        CompanyName:      Euro Plus d.o.o.
        ProductName:      Thermal Printers Driver
        InternalName:     ZDNDRV56.DLL
        OriginalFilename: ZDNDRV56.DLL
        ProductVersion:   5.1.16.6816
        FileVersion:      5.1.16.6816 built by: WinDDK
        FileDescription:  THERMAL Usermode Printer Driver
        LegalCopyright:   Copyright (C) 1997..2017 Euro Plus d.o.o.
```

**Comment by jkotas** [comment]
> Where do you see that the crash is from the printer driver?

The crash is a double-free bug. The crash is inside the Free method called at https://github.com/dotnet/runtime/blob/e55c908229e36f99a52745d4ee85316a0e8bb6a2/src/libraries/System.Drawing.Common/src/System/Drawing/Printing/PageSettings.cs#L171 in the latest dump and at https://github.com/dotnet/runtime/blob/e55c908229e36f99a52745d4ee85316a0e8bb6a2/src/libraries/System.Drawing.Common/src/System/Drawing/Printing/PageSettings.cs#L240 in the previous dump.

The memory was allocated here https://github.com/dotnet/runtime/blob/691cf0f448781df1379278943825a44c7120db75/src/libraries/System.Drawing.Common/src/System/Drawing/Printing/PrinterSettings.cs#L831 and the path from here to the place where it is freed looks pretty straightforward.

It means that the double-free was likely introduced by some other code loaded in the process. The uncommon 3rd party printer driver is likely suspect since we have not received a report like this from anybody else.

**Comment by jogibear9988** [comment]
And is it also not possible to print out the stacktrace to the Eventlog?
With this it would also be possible to find the root of the error without a Dump.

**Comment by jkotas** [comment]
> And is it also not possible to print out the stacktrace to the Eventlog?

Windows terminates the process immediately when it detects unmanaged heap corruption. I do not believe there is a way to customize what happens on unmanaged heap corruption. And even if there was a way, the code to handle the unmanaged heap corruption would have to avoid allocating any memory from unmanaged heap that is very severe restriction. It would be very hard to implement printing of the stack trace into eventlog with this restriction.

**Comment by ViktorHofer** [comment]
cc @JeremyKuhne

**Comment by jogibear9988** [comment]
My Problem now, I found the issue in our code. We created our Databaselayer with every Call. The DatabaseLayer should be a Singelton, only our Connection should be created on each access. But this should only be a Perfomance issue, not a application crashing one, cause it does only uses Managed Code. But the Memory usage increases (but it don't know why, it's only in unmaged), and after a time the app crashes. After my fix now (that the databselayer is only created once), it works correctly.
But I still would like to know why it crashes. I tried to build a minimal reproducible sample, but I could not recreate.
I could only recreate in our whole app. If someone is interested to see the error in the complete app, I could upload a sample, wich would crash (but it still would take an hour or so)
